src/clj/athens/self_hosted/web/persistence.clj

```diff
@@ -0,0 +1,102 @@
+(ns athens.self-hosted.web.persistence
+  (:refer-clojure :exclude [load list])
+  (:require
+    [clojure.data.json :as json]
+    [clojure.java.io :as io]
+    [clojure.string :as str]
+    [datascript.core :as d]))
+
+
+(def extension ".json")
+
+
+(defn- is-persisted-file?
+  [path]
+  (and (-> path io/file .isFile)
+       (str/ends-with? path extension)))
+
+
+(defn- id->path
+  [persist-base-path id]
+  (->> [id extension]
+       str/join
+       (io/file persist-base-path)
+       .getAbsolutePath))
+
+
+(defn- path->id
+  [path]
+  (if (is-persisted-file? path)
+    (-> path io/file .getName (str/replace extension ""))
+    (throw (ex-info "Path is not for a persisted file" {:path path}))))
+
+
+(defn- list
+  [persist-base-path]
+  (->> persist-base-path
+       io/file
+       .listFiles
+       (map #(.getAbsolutePath %))
+       (filter is-persisted-file?)))
+
+
+(defn- delete-others!
+  [persist-base-path id]
+  (->> (list persist-base-path)
+       (remove #{(id->path persist-base-path id)})
+       (run! io/delete-file)))
+
+
+(defn save!
+  [persist-base-path db id]
+  (let [path (id->path persist-base-path id)]
+    (io/make-parents path)
+    (->> db
+         d/serializable
+         json/write-str
+         (spit path))
+    (delete-others! persist-base-path id)
+    path))
+
+
+(defn load
+  [persist-base-path]
+  (when-some [path (-> (list persist-base-path) first)]
```

Comment:
So when there would be multiple "snapshot" files, we'd just take first one?

Reply:
Yes, with no sorting or anything. The only case I think this can happen, given that `save!` will delete other snapshots, is if the process crashes after `save!` makes a new snapshot but before deleting the others.

If this happened multiple times, for whatever reason, then you'd end up with multiple snapshots. But the next time `save!` finishes successfully, all others will be deleted.

I think this is a pretty solid failure mode. It's even safe with multiple processes. Wink wink.